### PR TITLE
Add Automation script to shard a pbf using AWS

### PR DESCRIPTION
**Pull request by atiannicelli**
### Description:

Add scripts and README to shard a pbf URL using AWS EC2 in preparation to generate atlas files. The README aded gives detailed descriptions of the scripts and how to use them.

### Potential Impact:

This will not impact any atlas-generation code. The scrip is meant to accompany the atlas-generator to provide AWS automaton support.

### Unit Test Approach:

No unit tests were added for this script. 

### Test Results:

I have tested this script by executing it against our AWS configuration. I successfully sharded 2 different planet files (this took ~4 hours each). I have also executed this script on a number of smaller pbf URLs and verified that the output of these runs were the same as manually produced sharding runs.

------

In doubt: [Contributing Guidelines](https://github.com/osmlab/atlas/blob/dev/CONTRIBUTING.md)

### CODE REVIEW

File: scripts/cloudPBFShardControl/cloudPBFShardControl.py [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+                logger.debug(
+                    "Waiting for systemd on EC2 instance to complete initialization..."
+                )
+                time.sleep(6)
+                continue
+            return
+        finish("Timeout while waiting for EC2 instance to be ready", -1)
+
+
+def parse_args(cloudctl):
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/cloudPBFShardControl.py:424]:
```suggestion
def parse_args(cloudctl: CloudPBFShardControl):
```

File: scripts/cloudPBFShardControl/cloudPBFShardControl.py [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+        "corresponding pem file must be located in your local '~/.ssh/' "
+        "directory and should be a pem file. See the following URL for "
+        "instructions on creating a key: "
+        "https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html. "
+        "(e.g. `--key=aws-key`)",
+    )
+    parser_prep.add_argument(
+        "-i", "--id", help="ID - Indicates the ID of an existing VM instance to use"
+    )
+    parser_prep.set_defaults(func=CloudPBFShardControl.prep)
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/cloudPBFShardControl.py:489]:
```suggestion
    parser_prep.set_defaults(func= cloudctl.prep)
```

File: scripts/cloudPBFShardControl/cloudPBFShardControl.py [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+        "(e.g. '--pbf=https://download.geofabrik.de/central-america-latest.osm.pbf')",
+    )
+    parser_shard.add_argument(
+        "-P",
+        "--processes",
+        type=int,
+        help="processes - The number of parallel osmium processes to start "
+        "(Default: {})".format(cloudctl.processes),
+    )
+    parser_shard.set_defaults(func=CloudPBFShardControl.shard)
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/cloudPBFShardControl.py:529]:
```suggestion
    parser_shard.set_defaults(func= cloudctl.shard)
```

File: scripts/cloudPBFShardControl/cloudPBFShardControl.py [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+        "corresponding pem file must be located in your local '~/.ssh/' "
+        "directory and should be a pem file. See the following URL for "
+        "instructions on creating a key: "
+        "https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html. "
+        "(e.g. `--key=aws-key`)",
+    )
+    parser_sync.add_argument(
+        "-o", "--out", required=True, help="Out - The S3 Output directory"
+    )
+    parser_sync.set_defaults(func=CloudPBFShardControl.sync)
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/cloudPBFShardControl.py:555]:
```suggestion
    parser_sync.set_defaults(func= cloudctl.sync)
```

File: scripts/cloudPBFShardControl/cloudPBFShardControl.py [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+        required=True,
+        help="KEY - Instance key name to use to login to instance. This key "
+        "is expected to be the same name as the key as defined by AWS and the "
+        "corresponding pem file must be located in your local '~/.ssh/' "
+        "directory and should be a pem file. See the following URL for "
+        "instructions on creating a key: "
+        "https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html. "
+        "(e.g. `--key=aws-key`)",
+    )
+    parser_clean.set_defaults(func=CloudPBFShardControl.clean)
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/cloudPBFShardControl.py:576]:
```suggestion
    parser_clean.set_defaults(func= cloudctl.clean)
```

File: scripts/cloudPBFShardControl/cloudPBFShardControl.py [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+        # awsCredentials = os.path.join(os.environ.get("HOME"), ".aws/credentials")
+        # logger.info(
+        #     "Pushing AWS config({}) and credentials({}) to EC2 instance.".format(
+        #         awsConfig, awsCredentials
+        #     )
+        # )
+        # remoteDir = "/home/ubuntu/.aws/"
+        # if self.ssh_cmd("mkdir -p {}".format(remoteDir)):
+        #     finish("Unable to create directory", -1)
+        # self.put_files([awsConfig, awsCredentials], remoteDir)
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/cloudPBFShardControl.py:175]: is this still useful? maybe remove if not

**Review by zlavergne — CHANGES_REQUESTED** [review]
Just a few minor things but overall it looks good. I can give it a more thorough pass if needed

File: scripts/cloudPBFShardControl/cloudPBFShardControl.py [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+    if hasattr(args, "out") and args.out:
+        cloudctl.s3Folder = args.out
+    if hasattr(args, "pbf") and args.pbf:
+        cloudctl.pbfURL = args.pbf
+    if hasattr(args, "id") and args.id:
+        cloudctl.instanceId = args.id
+        cloudctl.get_instance_info()
+    if args.version:
+        logger.critical("This is version {0}.".format(VERSION))
+        finish()
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/cloudPBFShardControl.py:594]: I thought I had put a comment on this in here but I didn't see it...

Something to consider with these checks

File: scripts/cloudPBFShardControl/pbfShardGenerator.py [hunk 8]
```diff
@@ -0,0 +1,715 @@
...
+        sharder.maxShardPerStep = args.max_shard_per_step
+    if hasattr(args, "max_shard_per_config") and args.max_shard_per_config:
+        sharder.maxShardPerConfig = args.max_shard_per_config
+    if hasattr(args, "out") and args.out:
+        sharder.s3Folder = args.out
+    if hasattr(args, "pbf") and args.pbf:
+        sharder.pbfURL = args.pbf
+    if hasattr(args, "processes") and args.processes:
+        sharder.maxOsmiumProcesses = args.processes
+    if args.version:
```
Comment (zlavergne) [inline scripts/cloudPBFShardControl/pbfShardGenerator.py:699]: Same convo as above about the conditional checks

**Review by matthieun — COMMENTED** [review]
This sounds like a great idea! Maybe @lucaspcram can take a look at the python code?

Just one nit, the sharding tree which is checked in... Is it for testing? If yes maybe a smaller one could do? Otherwise we'd rather have the user provide it, so we do not have to check it in. We have checked in large files in the past and regret it ever since!

**Comment by atiannicelli** [comment]
thanks @matthieun , good comment. I will refactor a bit to fetch the sharding tree file from S3 like the other scripts do and remove the sharding tree file from this commit.

**Review by lucaspcram — APPROVED** [review]
The python looks good to me!

**Review by jwpgage — APPROVED** [review]
Thanks for these scripts @atiannicelli!

**Review by matthieun — APPROVED** [review]
Nice!

**Pull request merged**